networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/peers/Eth2PeerSelectionStrategy.java

```diff
@@ -38,54 +47,149 @@
+        targetPeerCountRange.getRandomlySelectedPeersToAdd(randomlySelectedPeerCount);
+
     final int peersRequiredForSubnets = peerSubnetSubscriptions.getSubscribersRequired();
-    final int maxAttempts = Math.max(peersRequiredForPeerCount, peersRequiredForSubnets);
-    LOG.trace("Connecting to up to {} known peers", maxAttempts);
-    if (maxAttempts == 0) {
+    final int scoreBasedPeersToAdd =
+        Math.max(peersRequiredForPeerCount - randomlySelectedPeerCount, peersRequiredForSubnets);
+    LOG.trace(
+        "Connecting to up to {} known peers", scoreBasedPeersToAdd + randomlySelectedPeersToAdd);
+    if (scoreBasedPeersToAdd == 0 && randomlySelectedPeersToAdd == 0) {
       return emptyList();
     }
+
+    final List<DiscoveryPeer> allCandidatePeers = new ArrayList<>(candidates.get());
+    final List<PeerAddress> selectedPeers = new ArrayList<>();
+
+    if (randomlySelectedPeersToAdd > 0) {
+      selectedPeers.addAll(
+          selectAndRemoveRandomPeers(
+              network, peerSources, randomlySelectedPeersToAdd, allCandidatePeers));
+    }
+
+    if (scoreBasedPeersToAdd > 0) {
+      selectedPeers.addAll(
+          selectPeersByScore(
+              network, peerSubnetSubscriptions, scoreBasedPeersToAdd, allCandidatePeers));
+    }
+    return unmodifiableList(selectedPeers); // Unmodifiable to make errorprone happy
+  }
+
+  private List<PeerAddress> selectAndRemoveRandomPeers(
+      final P2PNetwork<?> network,
+      final PeerSources peerSources,
+      final int randomlySelectedPeersToAdd,
+      final List<DiscoveryPeer> allCandidatePeers) {
+    final List<PeerAddress> selectedPeers = new ArrayList<>();
+    shuffler.shuffle(allCandidatePeers);
+    while (!allCandidatePeers.isEmpty() && selectedPeers.size() < randomlySelectedPeersToAdd) {
+      final DiscoveryPeer candidate = allCandidatePeers.remove(0);
+      checkCandidate(candidate, network)
+          .ifPresent(
+              peerAddress -> {
+                peerSources.recordPeerSource(peerAddress.getId(), RANDOMLY_SELECTED);
+                selectedPeers.add(peerAddress);
+              });
+    }
+    return selectedPeers;
+  }
+
+  private List<PeerAddress> selectPeersByScore(
+      final P2PNetwork<?> network,
+      final PeerSubnetSubscriptions peerSubnetSubscriptions,
+      final int scoreBasedPeersToAdd,
+      final List<DiscoveryPeer> allCandidatePeers) {
+    final List<PeerAddress> selectedPeers = new ArrayList<>();
     final PeerScorer peerScorer = peerSubnetSubscriptions.createScorer();
-    return candidates.get().stream()
+    allCandidatePeers.stream()
         .sorted(
             Comparator.comparing((Function<DiscoveryPeer, Integer>) peerScorer::scoreCandidatePeer)
                 .reversed())
-        .map(network::createPeerAddress)
+        .flatMap(candidate -> checkCandidate(candidate, network).stream())
+        .limit(scoreBasedPeersToAdd)
+        .forEach(selectedPeers::add);
+    return selectedPeers;
```

Comment:
Am I missing the place where we record peer as `SELECTED_BY_SCORE`?
```java
peerSources.recordPeerSource(peer, SELECTED_BY_SCORE);
```

Reply:
`SELECTED_BY_SCORE` is the default category used if there's no entry so we only ever record peers as being static or randomly selected.  I've renamed `PeerSource` to `PeerPools` to make this clearer.